### PR TITLE
feat(taikoon): migrate Taikoon NFT smart contracts here

### PR DESCRIPTION
**Pull request by bearni95**
- [x] Copy the code over
- [x] Make the tests run
- [x] Make the scripts run

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,15 @@
+[submodule "packages/taikoon/lib/forge-std"]
```
Comment (dantaik) [inline .gitmodules:1]: I don't think we need this file.

File: packages/taikoon/data/script/add-images-ipfs.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+// Helper function to form the metadata JSON object
+function populateNFTMetadata(name, description, CID) {
+    return {
+        name,
+        description,
+        image: CID,
+    }
+}
+
+async function main() {
```
[the remark this answers is not in the document]
Comment (bearni95) [reply packages/taikoon/data/script/add-images-ipfs.js]: Made subdirs under the root `script` for `js` and `sol` separate scripts

File: packages/taikoon/contracts/MerkleWhitelist.sol [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+pragma solidity 0.8.24;
+
+import { MerkleProof } from "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+import { ContextUpgradeable } from
+    "@openzeppelin/contracts-upgradeable/utils/ContextUpgradeable.sol";
+
+/// @title MerkleWhitelist
+/// @dev Merkle Tree Whitelist
+/// @custom:security-contact security@taiko.xyz
+contract MerkleWhitelist is ContextUpgradeable {
```
Comment (dantaik) [inline packages/taikoon/contracts/MerkleWhitelist.sol:11]: for upgradable contract, we need to have a constructor like this:


```
    /// @custom:oz-upgrades-unsafe-allow constructor
    constructor() {
        _disableInitializers();
    }
```

**Comment by adaki2004** [comment]
> pnpm install



> Screenshot 2024-04-25 at 22 18 12
> 
> @adaki2004 do you see these errors after "pnpm install && pnpm test"?

I can see the same error when i run `forge test` (after pnpm install). 

(I cannot run tho `pnpm install && pnpm test` the exact smae way you copied here, because it throws some other shit. But when i run the forge test, it shows me these errors, yes).

**Review by dantaik — APPROVED** [review]

**Comment by dantaik** [comment]
The issue is with the library. Maybe it's not compatible with the compiler or OZ libraries we use.
Please remove it and use ` proxy = address(new ERC1967Proxy(new SomeContract(), data));` directly.

**Review by davidtaikocha — APPROVED** [review]

**Pull request merged**